tests/integration/utils.sh

```diff
@@ -157,6 +159,177 @@ function teardown_single_container()
+        image_exists "${dangling_img}"
+        if [[ "$?" -ne 0 ]]; then
+          continue
+        fi
+
+        # We check if the dangling image have layers which are exact prefixes of
+        # the current image we are trying to delete.
+        dangling_img_layer_list=$(image_inspect -f '{{.RootFS.Layers}}' "${dangling_img}" | tr -d '[]')
+        grep --perl-regexp "^${dangling_img_layer_list}" <<< "${img_layer_list}" > /dev/null
+
+        # Delete the dangling image if that is the case.
+        if [[ "$?" -eq 0 ]]; then
+          # Dangling images are always force-removed.
+          image_rm --force "${dangling_img}"
+        fi
+      done <<< "$(image_ls --all --quiet --filter 'intermediate=true')"
+
+      # Remove main image.
+      image_rm --force "${img}"
+      if [[ "$?" -ne 0 ]]; then
+        complain "Failed to remove ${img} image."
+      fi
+    fi
+
+    # Remove the base image for the distro.
+    img="${CONTAINER_BASE_IMAGE}/${distro}"
+    image_exists "${img}"
+    if [[ "$?" -eq 0 ]]; then
+      say "Removing podman ${img} image."
+
+      if [[ -n "$flag" ]]; then
+        image_rm "$flag" "$img"
+      else
+        image_rm "$img"
+      fi
+    fi
+  done
+}
+
+# Check if the given image exists.
+#
+# @image    The image name or id.
+function image_exists()
+{
+  podman image exists "$1"
+}
+
+# Remove the given images.
+#
+# @flags    Optional flags to be passed to podman.
+#           If it is '-f' or '--force', then it should be the ONLY flag.
+# @images   Image names or ids.
+function image_rm()
+{
+  local force_remove=0
+  local img
+
+  if [[ "$1" == "-f" || "$1" == "--force" ]]; then
+    force_remove=1
+    shift
+  fi
+
+  if [[ $# -lt 1 ]]; then
+    complain "(${LINENO}): no image provided to be removed."
+    return 1
+  fi
+
+  # shellcheck disable=SC2068
+  podman image rm $@ > /dev/null 2>&1
+
+  # command succeed.
+  if [[ "$?" -eq 0 ]]; then
+    return 0
+  fi
+
+  # comman failed and we should not force-attempt it.
+  if [[ "$force_remove" == 0 ]]; then
+    complain "(${LINENO}): kw failed to execute \`podman image rm ${*}\`. Consider using --force flag."
+    return 1
+  fi
+
+  # We try it again, this time forcing the removal of the images.
+  # WARNING: If the first argument is '-f', we suppose all the other ones are the image names.
+  while [[ $# -gt 0 ]]; do
+    img="$1"
+    shift
+
+    # Make sure the image exists.
+    podman image exists "$img"
+    if [[ "$?" -ne 0 ]]; then
+      complain "(${LINENO}) provided image '${img}' does not exist."
+      continue
+    fi
+
+    # Force remove all containers that depend on that image, one by one.
+    (podman container ls --all --quiet --filter ancestor="$img" |
+      xargs -n1 podman container rm --force --time 0) > /dev/null 2>&1
+
+    # Force remove the image now.
+    podman image rm --force "$img" > /dev/null 2>&1
+  done
+}
+
+# List the given images.
+#
+# @flags    Optional flags to be passed to podman.
```

Comment:
FWIU, this `flags` argument are podman options that will be passed through `$@` (along with the `images`) to the podman command. Instead of separating the documentation of these two arguments, I think it is better to describe them as one (all arguments passed to the function), and explicitly mentioning that it is composed of (first) podman options and (second) image names/IDs.